CEP-21 review

blerer opened the pull request:
The goal of this PR is to allow to collect the review/comments for the CEP-21 branch while allowing the addition of javadoc or minor code changes by the reviewers.
This will allow the reviewers to quickly benefit from the improvements in documentation to improve the understanding of the code.

ifesdjeen committed:
[CEP-21] Improve seedlist inspection at startup

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18404

beobal committed:
[CEP-21] Add distributed metadata log

Members of the ClusterMetadataService (CMS) replicate the global log
table, and are responsible for linearizing inserts into the the log.
Log entries contain transformations which are applied to ClusterMetadata
in the prescribed order. Log entries are replicated to non-CMS members
only after being ordered and inserted into the log.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Move schema management to TCM

Schema itself is a component of ClusterMetadata and all DDL updates are
applied by the CMS inserting a log entry containing a schema
transformation. As the log entries are disseminated around the cluster,
each peer applies the transformation to its local ClusterMetadata,
enacting the schema change. This entails some changes to the way the
database objects represented in schema are intialised (db objects refers
to classes like Keyspace, ColumnFamilyStore, etc).

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Include current epoch in internode header

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Add rudimentary cluster membership to TCM

Adds a new Directory component to ClusterMetadata to manage member
identity, state location and addressing. This duplicates some of the
functions of TokenMetadata, Topology et al but with updates performed
consistently via the global log. Although it isn't actually used for
anything yet it is a prerequisite for managing data ownership through
TCM, which will eventually replace TokenMetadata completely.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Add basics of ownership and data placement

Introduce new classes for representing placement of data ranges on
replicas, along with the movement of data via transitions from one
placement to the next. Eventually, these placements will be statically
calculated in response to events with alter either the topology of the
cluster (i.e. adding/removing/moving nodes) or the replication profile
of the data itself (i.e. creating/altering keyspaces). These triggering
events will be distributed and enacted consistently using the global log.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Produce placements equivalent to current replication strategies

Minimal modifications to AbstractReplicationStrategy implementations to
support the production of DataPlacements using ClusterMetadata while
retaining calculateNaturalReplicas. Also adds tests to compare the
output of both methods and assert their equivalence. Eventually, the
original implementations based on TokenMetadata will be retired and
will be retained in the test source to guard against regressions.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Initial transforms for modifying ownership

Introduces transformations to modify ClusterMetadata with the affect of
modifying data ownership and placement i.e. join, replace, move &
decommission. These operations do not simply modify metadata however,
so simple atomic updates to cluster metadata are not sufficient.
Streaming data in and out of nodes must also occur and obviously read
and write operations must continue whilst this is in progress. These
operations then are performed in phases, planned in advance and properly
linearized using the global log. e.g. to join an new node, the full set
of phased range movements required is calculated, generating an
actionable plan which is then serialised into ClusterMetadata.
Concurrent operations are permitted as long as they only affect disjoint
token ranges, ensuring that concurrent range movements remain safe and
cluster invariants are preserved at all times, including in the case of
the failure to complete any operation (i.e. failed bootstrap).

This commit only adds the transformations and supporting TCM components
(LockedRanges, InProgressSequences etc), the implementation of actually
performing the operations follows in subsequent commits.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Start to remove and deprecate gossip functionality

WIP commit (i.e. does not compile) beginning the process of removing
gossip as the source of truth regarding membership, ownership, topology
and data placement. This task will be split over mutiple commits.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Replace fake initial implementation of CMS membership

WIP commit (i.e. does not compile) replacing initial toy implementation
of CMS membership with proper implementation. Membership of the CMS is
determined by ownership of keyspaces with the META replication strategy
(more precisely, by being a member of the _read_ placements for meta
strategy keyspaces, a node is considered a member of the CMS).

Also implements more of the "real" [pre]initialization of the CMS, in
preparation for supporting upgrading a running cluster from a gossip
based system.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Major changes to StorageService to remove TMD and Gossip (1/7)

Part 1 of 7 commits applying the main changes migrate StorageService
away from managing state using TokenMetadata with updates propagated
using gossip.

This commit makes the initial bulk changes to StorageService itself and
thoroughly breaks compilation. Subsequent commits in the series fix the
main build before test code is updated later.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Placement and ownership related changes (2/7)

Part 2 of 7 moves most of the data placement/ownership code over to the
TCM structures.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Consistent read/write path (3/7)

Part 3 of 7 adds the ability to detect version mismatches between peers
on the read/write path and to handle such divergence. Lagging peers will
attempt to catch up from the CMS if the coordinator in a r/w operation
has seen newer metadata. Coordinators may fail writes if the cluster
metadata changes while the write is in flight, if the consistency level
can no longer be satisfied by the original replica plan.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Update ColumnFamilyStore (4/7)

Part 4 of 7 modifications to ColumnFamilyStore, mostly related to:
* ShardBoundaries
* DiskBoundaries

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Fix misc compilation errors (5/7)

Part 5 of 7 only compilation errors in non-test code are directly
related to TokenMetadata

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Remove TokenMetadata (6/7)

Part 6 of 7 Completely remove TokenMetadata, the intention is to bring
it back in a stripped down form, available to tests only, so we can
continue to verify equivalence between old and new code.

Test code is still extremely broken at this point, but non-test code is
buildable again, though almost certainly not actually runnable.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Restore operation mode reporting (7/7)

Part 7 of 7 brings StorageService.operationMode back into sync with
previous behaviour. Many external coordination tools depend on accessing
this state via JMX, so this is an important external interface.

This commit also adds a virtual version of the system.local table, as we
can fully construct the data for this from ClusterMetadata, meaning we no
longer the on-disk system table, though this is retained for now. In
future, more system tables can be virtualised (system.peers,
system_schema, etc).

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] New nodetool commands

Adds new nodetool commands to:
* list members of the CMS
* initiate a snapshot of ClusterMetadata via submitting a SealPeriod
  operation

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Add missing implementations to concurrent utils

Adds a handful of implementations to subclasses in the
org.apache.cassandra.utils.concurrent package

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Upgrade support

Following an upgrade, nodes in an existing cluster will enter a minimal
modification mode. In this state, the set of allowed cluster metadata
modifications is constrained to include only the addition, removal and
replacement of nodes, to allow failed hosts to be replaced during the
upgrade.

In this mode the CMS has no members and each peer maintains its
own ClusterMetadata independently. This metadata is intitialised at
startup from system tables and gossip is used to propagate the permitted
metadata changes.

When the operator is ready, one node is chosen for promotion to the initial
CMS, which is done manually via nodetool. At this point, the candidate node
will propose itself as the initial CMS and attempt to gain consensus from
the rest of the cluster. If successful, it verifies that all peers have an
identical view of cluster metadata and initialises the distributed log with
a snapshot of that metadata.

Once this process is complete all future cluster metadata updates are performed
via the CMS using the global log and reverting to the previous method of
metadata management is not supported. Further members can and should be added
to the CMS via the nodetool command.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Option to record thread creation stacktrace

Adds a property for use in tests and debugging  which preserves
the stacktrace of when a thread is created by NamedThreadFactory.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Update snitches to use Directory

Minimal changes to IEndpointSnitch implementations to have them pull
location info from Directory.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Modify CassandraDaemon

Alter CassandraDaemon intialization to accomodate TCM and replay of the
cluster metadata log. This is something of a WIP and there is clearly
scope to further clean up this part of the code.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Test / build config changes

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Bulk update of unit and dtests

Updates the existing unit and and dtests to work with TCM. In the vast
majority of cases, this just means changes to initialization or to
slightly updated method signatures.

In CEP-21 generally, the intention has been not to modify existing
public interfaces at all and to limit any changes to code on the
boundaries of internal subsystems. In addition, care has been taken to
only make minimal modifications to existing tests, and to preserve their
invariants. So although this commit is fairly large in terms of
number of files, it's changes should be semantically quite light.

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

ifesdjeen committed:
[CEP-21] Ensure that global log replication factor is maintained after decommission

patch by Alex Petrov; reviewed by Marcus Eriksson and Sam Tunnicliffe
for CASSANDRA-18416

ifesdjeen committed:
[CEP-21] Add debounce to log replay

patch by Alex Petrov; reviewed by Marcus Eriksson and Sam Tunnicliffe
for CASSANDRA-18402

ifesdjeen committed:
[CEP-21] Implement retries for log replay on CMS members

patch by Alex Petrov; reviewed by Marcus Eriksson and Sam Tunnicliffe
for CASSANDRA-18408

ifesdjeen committed:
[CEP-21] ReplicaPlan recomputation and stillAppliesTo implementation for Paxos

beobal committed:
TMP - use bundled version of harry

krummas committed:
[CEP-21] Correctly represent bootstrapping nodes in StorageService

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18409

krummas committed:
[CEP-21] Secondary indexes should not be rebuilt on restart

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18412

krummas committed:
[CEP-21] Re-enable stdout/sterr redirection at startup

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18414

beobal committed:
[CEP-21] Dereference TableMetadata in simple partition builder

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18418

krummas committed:
[CEP-21] Fix nodetool ring and effective ownership

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18410

beobal committed:
[CEP-21] During multi step operations, defer token map update until completion of final step

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18419

Commit:
[CEP-21] Always populate local gossip state at startup

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18403

beobal committed:
[CEP-21] Various fixes to in-jvm dtests

* JVMDTestTest
    schema change with stopped node, must not be CMS
* SSTableIdGenerationTest
    This test revolves around creating sstables with sequential ids, then
    restarting the node after enabling sstable uuids, creating more sstables
    and then counting the number of each type of sstable. It seems to rely
    on the local host id being unavailable when the node restarts, which
    causes the mutations in the commitlog, which have already been flushed
    to be replayed as the null local host id doesn't match the actual one in
    the sstable metadata. In turn this leads to an extra sstable being
    created after the restart, duplicating (some of) what was in the earlier
    sstables.
* MigrationCoordinatorTest
    deleted

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18405

beobal committed:
[CEP-21] Always use Paxos.v2 for global log reads/writes

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18406

krummas committed:
[CEP-21] Improve support for start/end tokens in nodetool rebuild

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18411

beobal committed:
[CEP-21] Remove paranoid check during SRPP

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18407

beobal committed:
[CEP-21] Add invalid routing exception

patch by Sam Tunnicliffe; reviewed by Alex Petrov and Marcus Eriksson
for CASSANDRA-18413

krummas committed:
[CEP-21] Fix (re)building MVs

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18415

beobal committed:
[CEP-21] Ensure that ClusterMetadata::forceEpoch keeps component epochs consistent

patch by Sam Tunnicliffe; reviewed by Marcus Eriksson and Alex Petrov
for CASSANDRA-18460

beobal committed:
[CEP-21] Ensure that SchemaTransformation impls correctly set TableMetadata epoch

patch by Sam Tunnicliffe; reviewed by Marcus Eriksson and Alex Petrov
for CASSANDRA-18457

beobal committed:
[CEP-21] During startup, don't open SSTables until local metadata log replay is complete

patch by Sam Tunnicliffe; reviewed by Marcus Eriksson and Alex Petrov
for CASSANDRA-18458

beobal committed:
[CEP-21] Rewrite o.a.c.distributed.test.SchemaTest

patch by Sam Tunnicliffe; reviewed by Marcus Eriksson and Alex Petrov
for CASSANDRA-18459

Commit:
[CEP-21] Implement multi-dc placement simulator for NTS

patch by Alex Petrov; reviewed by Marcus Eriksson and Sam Tunnicliffe
for CASSANDRA-18417

krummas committed:
[CEP-21] During startup request replay from CMS asynchronously

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18456

krummas committed:
[CEP-21] Avoid NPE when getting dc/rack for not yet registered endpoints

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18461

krummas committed:
[CEP-21] Fix tools tests

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18462

beobal committed:
[CEP-21] Remove some of the remaining vestiges of schema pulling

beobal committed:
[CEP-21] Bring back (most of) ring/BootstrapTest

beobal committed:
[CEP-21] Remove redundant or outdated tests

beobal committed:
[CEP-21] Make cleanup safe to run during range movements

beobal committed:
[CEP-21] Don't include system_distributed in initial schema

beobal committed:
[CEP-21] Fix eclipse warning

krummas committed:
[CEP-21] Reinstate client notifications for joining/leaving/moving nodes

patch by Marcus Eriksson; reviewed by Alex Petrov and Sam Tunnicliffe
for CASSANDRA-18463

Commit:
[CEP-21] Simulate NTS

Commit:
[CEP-21] Add exception code to commit result if rejected

DDL statements submitted by clients may return one of several different error responses

Commit:
[CEP-21] Add vtable for metadata log

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] Fix o.a.c.locator.NetworkTopologyStrategyTest

beobal committed:
[CEP-21] Re-enable startup checks on non-test initialization

Commit:
[CEP-21] Unify streaming: make all operations use explicit ranges for streaming

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] DC and RF aware acknowledgments for multistep operations

Commit:
[CEP-21] Gossip fixes

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

beobal committed:
[CEP-21] fix o.a.c.dht.tokenallocator.TokenAllocationTest

Commit:
Fix ByteBuddyExamplesTest

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

beobal committed:
[CEP-21] fix o.a.c.tools.SSTableExportSchemaLoadingTest

Commit:
[CEP-21] Fix replacement tests

Commit:
[CEP-21] Fix paxos self execution tests

beobal committed:
[CEP-21] Enforce data ownership checks

* in read and write verb handlers
* in repair message verb handler
* during streaming

beobal committed:
[CEP-21] Fix third party snitch tests

beobal committed:
[CEP-21] Fix o.a.c.locator.ReplicationFactorTest

beobal committed:
[CEP-21] Fix o.a.c.db.compaction.CompactionIteratorTest

beobal committed:
[CEP-21] Simplify check for whether DROP COMPACT STORAGE is permitted

beobal committed:
[CEP-21] Move local node registration from CQLTester -> ServerTestUtils

beobal committed:
[CEP-21] Move CQLTester initialisation from static block to @BeforeClass

beobal committed:
[CEP-21] Fix/remove fixture setup in several CQLTester subclasses

beobal committed:
[CEP-21] Update tests which require the local node not to be joined by default

beobal committed:
[CEP-21] Update tests which require the local node not to be joined by default

beobal committed:
[CEP-21] Fix nodetool cache invalidation tests

ifesdjeen committed:
[CEP-21] Support nodetool assasinate

beobal committed:
[CEP-21] Remove superseded join/move/leave/remove unit tests

These are better covered by the new fuzzing in-jvm dtests and
MetadataPlacementSimulatorTest

ifesdjeen committed:
[CEP-21] Reduce number of data harry writes to reduce flakiness

Commit:
[CEP-21] Fix reprepare tests

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] Python dtest fixes
* maybe fix hintedhandoff test

Commit:
[CEP-21] In progress sequence updates

* Protect sequenced operations from out-of-order and repeated execution (#3437)
* Improve sequence re-discovery
* Improve reliability of InProgressSequences
* Fix retries in AddToCMS. Remove retries from sequence commits, since they are pointless, keep retries for paxos repair

beobal committed:
[CEP-21] Don't invalidate prepared stmt cache on every schema change

beobal committed:
[CEP-21] Fix o.a.c.cql3.BatchTest

beobal committed:
[CEP-21] Fix o.a.c.cql3.Memtable*Test

ifesdjeen committed:
[CEP-21] Fix flaky consistentleave test

beobal committed:
[CEP-21] Fix o.a.c.cql3.OutOfSpaceTest

beobal committed:
[CEP-21] Fix o.a.c.cql3.statements.DescribeStatementTest

beobal committed:
[CEP-21] Fix o.a.c.cql3.validation.entities.SecondaryIndexTest

beobal committed:
[CEP-21] Fix o.a.c.cql3.validation.operations.AlterNTSTest

beobal committed:
[CEP-21] Fix o.a.c.cql3.validation.operations.DropRecreateAndRestoreTest

beobal committed:
[CEP-21] Fix o.a.c.db.DiskBoundaryManagerTest

beobal committed:
[CEP-21] Fix o.a.c.db.KeyspaceTest

beobal committed:
[CEP-21] Fix o.a.c.db.TopPartitionTrackerTest

beobal committed:
[CEP-21] Fix o.a.c.db.compaction.CompactionStrategyManagerBoundaryReloadTest

beobal committed:
[CEP-21] fix o.a.c.db.rows.ThrottledUnfilteredIteratorTest

beobal committed:
[CEP-21] Fix o.a.c.db.virtual.JmxPermissionsCacheKeysTableTest

beobal committed:
[CEP-21] Fix o.a.c.db.virtual.LocalRepairTablesTest

beobal committed:
[CEP-21] Fix o.a.c.db.virtual.StreamingVirtualTableTest

beobal committed:
[CEP-21] Fix o.a.c.service.LegacyAuthFailTest

beobal committed:
[CEP-21] Fix o.a.c.service.reads.range.TokenUpdater

beobal committed:
[CEP-21] Fix o.a.c.tools.JMXCompatibilityTest

beobal committed:
[CEP-21] Fix o.a.c.tools.StandaloneSplitterWithCQLTesterTest

beobal committed:
[CEP-21] Tests no longer need to manually create system_auth

Commit:
[CEP-21] Fix IPMembershipTest

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] Fix MoveTest

Commit:
[CEP-21] Fix assassinate dtests

beobal committed:
[CEP-21] Adjust to new Register checks

Test setup needs to handle the improved validation when executing Register
that was introduced to support nodes changing IP (in
b940e18d8983e12bda72ca0f27982616193f6f25).

In test code, this is part of an ongoing effort to rationalise the
various test setup code in the project and there is still more to do.

beobal committed:
[CEP-21] Adjust test for GossipInfo now STATUS is removed

beobal committed:
[CEP-21] Fix o.a.c.service.JoinTokenRingTest

beobal committed:
[CEP-21] Fix o.a.c.service.PartitionDenylistTest

beobal committed:
[CEP-21] Fix o.a.c.service.StorageServiceServerM3PTest

beobal committed:
[CEP-21] Additional tests which no longer need to manually create system_auth

beobal committed:
[CEP-21] Fix o.a.c.service.StorageProxyTest

beobal committed:
[CEP-21] Fix o.a.c.tcm.RemoteProcessorTest

beobal committed:
[CEP-21] Fix o.a.c.tcm.ClusterMetadataTransformationTest

beobal committed:
[CEP-21] Fix o.a.c.gms.GossiperTest

beobal committed:
[CEP-21] Rewrite o.a.c.service.StorageServiceServerTest

beobal committed:
[CEP-21] Fix o.a.c.schema.SchemaKeyspaceTest

beobal committed:
[CEP-21] Remove o.a.c.schema.RemoveWithoutDroppingTest

beobal committed:
[CEP-21] Fix o.a.c.db.compaction.writers.CompactionAwareWriterTest

beobal committed:
[CEP-21] Fix o.a.c.db.VerifyTest

beobal committed:
[CEP-21] Fix o.a.c.db.ColumnFamilyMetricTest

beobal committed:
[CEP-21] Fix o.a.c.db.CleanupTransientTest

beobal committed:
[CEP-21] Remove o.a.c.gms.PendingRangeCalculatorServiceTest

beobal committed:
[CEP-21] Remove o.a.c.gms.ExpireEndpointTest.java (CMS updates now remove from gossip)

beobal committed:
[CEP-21] Fix o.a.c.cql3.KeyCacheCqlTest

beobal committed:
[CEP-21] Fix o.a.c.tcm.transformations.PrepareLeaveTest

beobal committed:
[CEP-21] Fix o.a.c.tcm.transformations.EventsMetadataTest

beobal committed:
[CEP-21] Fix o.a.c.locator.ReplicaPlansTest

beobal committed:
[CEP-21] Fix o.a.c.locator.DynamicEndpointSnitchTest

beobal committed:
[CEP-21] Allow Schema.instance to be initialized empty for client apps

beobal committed:
[CEP-21] Remove obsolete o.a.c.locator.ReplicationStrategyEndpointCacheTest

beobal committed:
[CEP-21] ClusterMetadata::writePlacementAllSettled handles bootstrapping nodes correctly

beobal committed:
[CEP-21] Fix o.a.c.tcm.RemoteProcessorTest

To be squashed with 01386aa42675c9a64cdbe297486116a000ada654

beobal committed:
[CEP-21] Fix o.a.c.db.CleanupTest

beobal committed:
[CEP-21] Restore previous behaviour when fetching effective ownership for a local system keyspace

beobal committed:
[CEP-21] Fix o.a.c.service.reads.repair.[BlockingReadRepairTest|DiagEventsBlockingReadRepairTest]

beobal committed:
[CEP-21] Fix o.a.c.db.view.ViewUtilsTest

beobal committed:
[CEP-21] Fix o.a.c.service.BootstrapTransientTest

beobal committed:
[CEP-21] Fix MockSchema to use supplied keyspace name

beobal committed:
[CEP-21] Syncronize processing of LocalLog in test code

beobal committed:
[CEP-21] Fix o.a.c.tcm.ownership.DeltaMapTest

beobal committed:
[CEP-21] Disable auto metadata snapshots during tests

beobal committed:
[CEP-21] Fix o.a.c.io.sstable.format.RangeAwareSSTableWriterTest

beobal committed:
[CEP-21] When deserializing partition update, allow if current epoch >= serialized epoch

beobal committed:
[CEP-21] Fix o.a.c.tools.TopPartitionsTest

beobal committed:
[CEP-21] Make CQLSSTableWriter exclusively a client utility

beobal committed:
[CEP-21] Fix o.a.c.streaming.StreamRateLimiterTest

beobal committed:
[CEP-21] Fix o.a.c.db.StreamRequestTest

Commit:
[CEP-21] Update bootstrap log messages to match trunk

beobal committed:
[CEP-21] Fix o.a.c.dht.BootstrapperTest

Commit:
[CEP-21] Various JVM Dtest fixes

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Marcus Eriksson <marcus_eriksson@apple.com>

Commit:
[CEP-21] More in-jvm dtest fixes

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] Switch nodeId from uuid to int

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Reenable write survey mode

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Disaster recovery

Co-authored-by: Sam Tunnicliffe <samt@apache.org>
Co-authored-by: Marcus Eriksson <marcus_eriksson@apple.com>
Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

beobal committed:
[CEP-21] Mark successful migration as complete on nodes which did not initiate it

beobal committed:
[CEP-21] Require Paxos V2 for cluster metadata log operations

beobal committed:
[CEP-21] Reenable starting without joining (-Dcassandra.join_ring=false)

Commit:
[CEP-21] Support catching up from peers

Commit:
[CEP-21] Update system.local host id on upgrades, and bounces to fix current broken cluster (#3496)

Commit:
[CEP-21] Transfer CMS membership before replacement (#3498)

Commit:
[CEP-21] avoid npe when replaying from snapshot and endpoint shows up first time as LEFT (#3499)

Commit:
[CEP-21] Add nodetool command to remove from CMS

* Require explicit intent if shrinking CMS to fewer than 3 members
* Add nodetool command to remove node from the CMS

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] getHostIdForEndpoint return null if unknown endpoint

Commit:
[CEP-21] Simulator harry integration

Commit:
[CEP-21] add more metrics

Commit:
[CEP-21] Extend nodetool describecms output (#3507)

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Issues found during upgrade testing

* [CEP-21] Never try to catch up to an epoch lower than Epoch.FIRST
* [CEP-21] Ensure hints written before upgrade can be delivered afterward

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Fix table metadata serialization

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Simplistic schema metadata diff

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

ifesdjeen committed:
[CEP-21] Avoid using batches in distributed metadata log keyspace

ifesdjeen committed:
[CEP-21] Add keyspace to read and write recompute

beobal committed:
[CEP-21] Don't warn about new system tables in StartupCheck

ifesdjeen committed:
[CEP-21] Fix batchlog consistency errors during epoch bumps

beobal committed:
[CEP-21] Exclude meta keyspace from TableMetrics::totalNonSystemTablesSize

ifesdjeen committed:
[CEP-21] Use harry for generating data for coordinator path test

beobal committed:
Circle config

Commit:
[CEP-21] Multi cms simulation, Deadlines for local processor, reworked retries for local and remote processor

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] Fix issue when catching up from snapshot and the pending log is non-consecutive

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] Escape infinite local log loop on replica mis-configuration

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

Commit:
[CEP-21] fix MutualTls tests

Commit:
[CEP-21] make ColumnFamilyStoreClientModeTest build

Commit:
[CEP-21] Create new auth tables, remove cidr constants for column names

Commit:
[CEP-21] Update index status fixes post-rebase

ifesdjeen committed:
[CEP-21] Add a script for simulation running

Commit:
[CEP-21] Avoid gossip deadlock by grabbing taskLock before runInGossipStageBlocking

Commit:
[CEP-21] Evict replaced endpoint from gossip

beobal committed:
[CEP-21] Preparing DDL statements should capture the CQL string

krummas committed:
rebase fixes

beobal committed:
[CEP-21] Validate schema alterations on coordinator

Also ensure that alter/create table will succeed when being processed
either by a CMS node committing the entry to the log, or a log follower
enacting a committed transformation. If the specified memtable config
is not present in local yaml, fall back to the default.

krummas committed:
rebase fix

krummas committed:
fix UnifiedCompactionStrategyTest

krummas committed:
more test fixes

Commit:
[CEP-21] Add metric for CMS membership

beobal committed:
[CEP-21] Don't trigger client warnings or guardrails when executing AlterSchema transformations

Only when a coordinator is preparing to submit an AlterSchemaStatement
to the CMS.

beobal committed:
[CEP-21] Work with Storage Compatiblility Mode

* Add epoch to payload of current epoch req/rsp
* Skip token validation checks if in storage compatibility mode

krummas committed:
storage_compat_mode: none

Commit:
[CEP-21] Add nextId to Directory serialization and make it possible to bump metadata serialization version

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Marcus Eriksson <marcus_eriksson@apple.com>

beobal committed:
[CEP-21] Remove redundant Keyspaces arg from SchemaTransformation::apply

beobal committed:
[CEP-21] Handle case where removenode requires no streaming to restore RF

Commit:
[CEP-21] Implement versioning for ranges

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>
Co-authored-by: Marcus Eriksson <marcus_eriksson@apple.com>

ifesdjeen committed:
[CEP-21] Fix flaky distributed log test. While it fails very infrequently on CI, it consistently fails locally.

Commit:
[CEP-21] Retry indefinitely for STARTUP messages.

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>

beobal committed:
[CEP-21] Remove LEFT peers from system tables and exclude them during startup

* Don't try to connect to them with StartupClusterConnectivityChecker
* Don't pre-emptively mark them as DOWN in Gossiper::waitToSettle

Commit:
[CEP-21] fix nodetool bootstrap resume

Co-authored-by: Sam Tunnicliffe <samt@apache.org>

krummas committed:
[CEP-21] serialize MemtableParams when writing TableParams

Commit:
[CEP-21] Implement replacement with same address

Co-authored-by: Alex Petrov <oleksandr.petrov@gmail.com>
Co-authored-by: Sam Tunnicliffe <samt@apache.org>

Commit:
[CEP-21] Upgrading a one node cluster to TCM fails attempting Gossip shadow round

krummas committed:
rebase

krummas committed:
[CEP-21] remove authsetup

krummas committed:
[CEP-21] fix cqlshlib tests

krummas committed:
update dtest repo for cci

krummas committed:
[CEP-21] CASSANDRA-18816 rebase fixes

krummas committed:
[CEP-21] fix GossiperTest - this test now matches trunk

blerer committed:
Add javadoc WIP

blerer committed:
Simplify the RecentlySealedPeriod logic

jacek-lewandowski committed:
Add JavaDoc - jacek - wip

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/ClusterMetadata.java:
```diff
@@ -0,0 +1,986 @@
+                    mutable.withLastModified(epoch);
+            }
+
+            if (schema != base.schema)
```
why do we compare all those things by references instead of values (like dumpDiff does)?

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/membership/Directory.java:
```diff
@@ -0,0 +1,749 @@
+import static org.apache.cassandra.db.TypeSizes.sizeof;
+import static org.apache.cassandra.tcm.membership.NodeVersion.CURRENT;
+
+public class Directory implements MetadataValue<Directory>
```
consider renaming this class - maybe `NodeDirectory`? or `NodeIndex`? or `Nodes`?

jacek-lewandowski commented on src/java/org/apache/cassandra/schema/DistributedSchema.java:
```diff
@@ -18,38 +18,261 @@
+        initializeKeyspaceInstances(prev, true);
+    }
+
+    public void initializeKeyspaceInstances(DistributedSchema prev, boolean loadSSTables)
```
I think that managing `Keyspace` instances does not belong to this class - this is immutable object and as such should not modify itself. Maybe this could be moved back to `Schema`?

blerer replied:
Yes, this part confused me too.

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/Period.java:
```diff
@@ -0,0 +1,211 @@
+ * Utility methods and classes to work with {@code Period}.
+ *
+ */
+public class Period
```
consider renaming this class, maybe `LogUtils`, `PeriodUtils`, `Periods`?

jacek-lewandowski commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -428,11 +455,16 @@ public static Prepared parseAndPrepare(String query, ClientState clientState, bo
+        // Set CQL string for AlterSchemaStatement as this is used to serialize the transformation
+        // in the cluster metadata log
+        if (statement instanceof AlterSchemaStatement)
+            ((AlterSchemaStatement)statement).setCql(query);
```
would it be possible to serialize schema transformation in more formalized way? I don't know if it is possible to have bound variables in alter schema statement, but if so, or it will be possible in the future, what would happen then?

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/log/Replication.java:
```diff
@@ -0,0 +1,281 @@
+/**
+ * A sorted set of log entries mainly used to replicate log entries within the cluster.
+ */
+public class Replication
```
consider renaming this class, some suggestions: LogEntries, LogEntryCollection...

ekaterinadimitrova2 replied:
Further ideas `ReplicationLogEntries`, `ReplicationSet`

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/log/SystemKeyspaceStorage.java:
```diff
@@ -0,0 +1,168 @@
+        }
+        catch (Throwable t)
+        {
+            logger.error("Could not persist the entry {} proceeding with in-memory commit.", entry, t);
```
no feedback to the caller - why? can we ignore the error and continue?

jacek-lewandowski commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -181,14 +193,15 @@ public enum Kind
 
     // performance hacks; TODO see if all are really necessary
     public final DataResource resource;
+    public TableMetadataRef ref;
```
what is the purpose of ref initialized in TableMetadata?

jacek-lewandowski commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -459,6 +468,45 @@ private SystemKeyspace()
           + "cfids set<uuid>, "
           + "PRIMARY KEY (parent_id))").build();
 
+    public static final TableMetadata LocalMetadataLog =
```
I'm not sure if I understand that correctly - there are four additional tables, but it looks like we need only one:
- we need to be able to store a transformation per period and epoch
- we need to store a snapshot per period
- we need to be able to go through transformations sorted by epochs
- we need to be able to find the latest period and the latest epoch

Given we only have LocalMetadataLog, which stores the ultimate cluster metadata snapshot in a static column, we can probably get rid of the snapshots table;

Old entries from the log should be evicted, so there should be little partitions representing periods. If so, we could fetch all period - min/max epoch tuples by performing a full table scan and memoizing the results. That would eliminate the need for the remaining two tables with sealed periods and the last sealed period. 

Even if the table turns out to be big, making full scan too costly, we can leverage that each sstable contains min/max clustering information in the statistics metadata. By duplicating the period column as the first clustering column, we could quickly select the relevant sstables by only inspecting their metadata?

Even if we need to store the last epoch and period, we could put them into `local` table.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/DistributedMetadataLogKeyspace.java:
```diff
@@ -0,0 +1,283 @@
+    }
+
+    @VisibleForTesting
+    public static void truncateLogState()
```
This method is not used anywhere.

ekaterinadimitrova2 commented on .build/parent-pom-template.xml:
```diff
@@ -336,6 +336,11 @@
+      <dependency>
+        <groupId>com.google.guava</groupId>
+        <artifactId>guava-testlib</artifactId>
+        <version>27.0-jre</version>
```
I suspect this should also b updated to 32.0.1 as guava, at least.

jacek-lewandowski replied:
yeah, perhaps use some internal property for guava version

ekaterinadimitrova2 commented on ci/harry_simulation.sh:
```diff
@@ -0,0 +1,96 @@
+        -XX:Tier4CompileThreshold=1000
+        -XX:ReservedCodeCacheSize=256M
+        -Xmx16G
+        -Xmx4G
```
We should keep either 16 or 4 for Xmx or maybe it was meant to be Xmx and Xms? In our JVM options we recommend both set to 4G

ekaterinadimitrova2 commented on ci/harry_simulation.sh:
```diff
@@ -0,0 +1,96 @@
+        -Xmx16G
+        -Xmx4G
+        -Xss384k
+        --add-exports java.base/jdk.internal.misc=ALL-UNNAMED
```
Have you considered pulling all these add-exports and add-opens from one place for all tests? They are copy-pasted, and only the Java 11 batch, which I suspect will quickly regress when we are updating JDKs.

ekaterinadimitrova2 commented on conf/cassandra.yaml:
```diff
@@ -1184,16 +1184,16 @@ range_request_timeout: 10000ms
 # Lowest acceptable value is 10 ms.
 # Min unit: ms
-write_request_timeout: 2000ms
+write_request_timeout: 10000ms
```
Was this change of default a mistake?
I can create a patch on confirmation to either revert it or update Config and the test suites config files.

ekaterinadimitrova2 commented on conf/cassandra.yaml:
```diff
@@ -1184,16 +1184,16 @@ range_request_timeout: 10000ms
 # Min unit: ms
-counter_write_request_timeout: 5000ms
+counter_write_request_timeout: 1000ms
 # How long a coordinator should continue to retry a CAS operation
```
Was this change of default a mistake?
I can create a patch on confirmation to either revert it or update Config and the test suites config files.

ekaterinadimitrova2 commented on conf/cassandra.yaml:
```diff
@@ -1184,16 +1184,16 @@ range_request_timeout: 10000ms
 # that contends with other proposals for the same row.
 # Lowest acceptable value is 10 ms.
 # Min unit: ms
-cas_contention_timeout: 1000ms
```
Was this change of default a mistake?
I can create a patch on confirmation to either revert it or update Config and the test suites config files.

ekaterinadimitrova2 commented on conf/cassandra.yaml:
```diff
@@ -2109,4 +2109,4 @@ drop_compact_storage_enabled: false
 #   a stable cluster. If a node from a previous version was started by accident we won't any longer toggle behaviors as when UPGRADING.
 #
-storage_compatibility_mode: CASSANDRA_4
+storage_compatibility_mode: NONE
```
Leaving this comment as a placeholder, as I see there is TODO in config, so we need to figure out this one.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -348,7 +348,7 @@ public enum CassandraRelevantProperties
     MEMTABLE_SHARD_COUNT("cassandra.memtable.shard.count"),
     MEMTABLE_TRIE_SIZE_LIMIT("cassandra.trie_size_limit_mb"),
-    MIGRATION_DELAY("cassandra.migration_delay_ms", "60000"),
+    METRICS_REPORTER_CONFIG_FILE("cassandra.metricsReporterConfigFile"),
```
I suspect this one was added back by mistake? We removed the metrics reporter recently.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -348,7 +348,7 @@ public enum CassandraRelevantProperties
     MEMTABLE_OVERHEAD_SIZE("cassandra.memtable.row_overhead_size", "-1"),
     MEMTABLE_SHARD_COUNT("cassandra.memtable.shard.count"),
     MEMTABLE_TRIE_SIZE_LIMIT("cassandra.trie_size_limit_mb"),
-    MIGRATION_DELAY("cassandra.migration_delay_ms", "60000"),
```
Shouldn't this be deprecated before removal?

krummas replied:
it has no effect anymore, and for users setting it cassandra will still start without issue

ekaterinadimitrova2 replied:
I am with you on this; it was just something people do in the community for one major release, even if the property has no effect. But on second thought, that makes more sense for the Config class than here. In this class, the Deprecated parameters can still be used. I just checked them.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -450,7 +450,6 @@ public enum CassandraRelevantProperties
     SAI_TEST_BALANCED_TREE_DEBUG_ENABLED("cassandra.sai.test.balanced_tree_debug_enabled", "false"),
     SAI_TEST_DISABLE_TIMEOUT("cassandra.sai.test.disable.timeout", "false"),
 
-    SCHEMA_PULL_INTERVAL_MS("cassandra.schema_pull_interval_ms", "60000"),
```
Shouldn't this be deprecated before removal?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -122,7 +122,7 @@ public static Set<String> splitCommaDelimited(String src)
     public CommitFailurePolicy commit_failure_policy = CommitFailurePolicy.stop;
 
-    public volatile boolean use_deterministic_table_id = false;
+    public volatile boolean use_deterministic_table_id = true;
```
Was this changed intentionally?

krummas replied:
temporarily yes, we'll post a patch that fixes deterministic table ids soon

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -171,6 +171,20 @@ public static Set<String> splitCommaDelimited(String src)
 
+    public volatile DurationSpec.LongMillisecondsBound cms_await_timeout = new DurationSpec.LongMillisecondsBound("120000ms");
+    public volatile int cms_default_max_retries = 10;
+    public volatile DurationSpec.IntMillisecondsBound cms_default_retry_backoff = new DurationSpec.IntMillisecondsBound("50ms");
```
I think we can remove the "default_" part, no? It is a configurable parameter whose default value is set, no? Same for `cms_default_max_retries`

ifesdjeen replied:
I think `default_` predicate makes sense here. I did think of calling it `start_` or `initial_` but it seems to be more confusing. We use the same term for indicating the protocol version nodes start negotiating by default.

ekaterinadimitrova2 replied:
A quick look shows me we have diverged from using `initial_` to `default_` in the past year in the `Config` class. Now, we are inconsistent. The precedent was not created in this patch; sadly, as it could be confusing to people, it happened. I leave it up to you; naming is hard in general.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -171,6 +171,20 @@ public static Set<String> splitCommaDelimited(String src)
 
     public volatile DurationSpec.LongMillisecondsBound stream_transfer_task_timeout = new DurationSpec.LongMillisecondsBound("12h");
 
+    public volatile DurationSpec.LongMillisecondsBound cms_await_timeout = new DurationSpec.LongMillisecondsBound("120000ms");
```
Out of curiosity, how were default values for all new parameters chosen? Methodology?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -4917,4 +4964,100 @@ public static RepairRetrySpec getRepairRetrySpec()
+        return conf.cms_default_max_retries;
+    }
+
+    public static void setCmsDefaultRetryMaxTries(int value)
```
This method seems to be not used and the name does not really match the parameter. Should I remove it?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/log/SystemKeyspaceStorage.java:
```diff
@@ -0,0 +1,168 @@
+    {
+        String query = String.format("SELECT epoch FROM %s.%s LIMIT 1", SchemaConstants.SYSTEM_KEYSPACE_NAME, NAME);
+
+        for (UntypedResultSet.Row row : executeInternal(query))
```
executeInternal can return null?

jacek-lewandowski replied:
I suppose it returns null when the statement does not return rows (for example, a schema update, or grant/revoke)

ekaterinadimitrova2 replied:
We need to take precautions with this loop.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/sequences/ProgressBarrier.java:
```diff
@@ -0,0 +1,568 @@
+    private static final Logger logger = LoggerFactory.getLogger(ProgressBarrier.class);
+    private static final ConsistencyLevel MIN_CL = DatabaseDescriptor.getProgressBarrierMinConsistencyLevel();
+    private static final ConsistencyLevel DEFAULT_CL = DatabaseDescriptor.getProgressBarrierDefaultConsistencyLevel();
+    private static final long TIMEOUT_MILLIS = DatabaseDescriptor.getProgressBarrierTimeout(TimeUnit.MILLISECONDS);
```
How was 1 hour chosen?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/transformations/CustomTransformation.java:
```diff
@@ -0,0 +1,264 @@
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.tcm.sequences.LockedRanges;
+
+public class CustomTransformation implements Transformation
```
Can you add a bit more information about this one, please?

ekaterinadimitrova2 committed:
Suggested tiny fixes
No CI was run as it is just small fixes and CI is having a bad day with this branch anyway

ifesdjeen commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -511,8 +511,9 @@ public enum CassandraRelevantProperties
+
     /**
-     * size of in-memory index of max epoch -> sealed period
+     * Maximum sized of the {@code RecentlySealedPeriods} in-memory index.
```
nit: sized -> size

ifesdjeen commented on src/java/org/apache/cassandra/schema/SchemaTransformation.java:
```diff
@@ -37,7 +37,7 @@ public interface SchemaTransformation
      * Apply a statement transformation to a schema snapshot.
      * <p>
-     * Implementing methods should be side-effect free (outside of throwing exceptions if the transformation cannot
+     * Implementing methods should be side effect free (outside of throwing exceptions if the transformation cannot
```
`side-effect` is also a correct spelling: https://www.collinsdictionary.com/dictionary/english/side-effect.

I would say `side-effect` but `side-effect free`.

ifesdjeen commented on src/java/org/apache/cassandra/tcm/ClusterMetadataService.java:
```diff
@@ -125,6 +125,10 @@ public static ClusterMetadataService instance()
     private final AtomicBoolean commitsPaused = new AtomicBoolean();
 
+    /**
+     * Returns the state of the {@code ClusteMetadataService}.
```
This comment seems to be tautological. Maybe add "returns _current_ state" at least? I would say it could be useful to specify what gets returned, too. Ie whether or not current cluster metadata service is local (ie member of CMS), remote (ie a non-CMS member of the cluster), gossip (ie a node that has not yet been upgraded to TCM), or the node is forced to reset the local metadata.

blerer replied:
I think that you answer the question that I was wondering. It was unclear to me what the different states means.
How does work the reset of the local metadata? 
As it is based on a property does it means that the node will stay on a reset state until it is restarted?

blerer committed:
Fix ClusterMetadataService.state javadoc

blerer committed:
Guaranty the ordering of entries within Replication instances

jacek-lewandowski committed:
fix ClusterMetadataService doc

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/ClusterMetadata.java:
```diff
@@ -0,0 +1,986 @@
+    public final Epoch epoch;
+    public final long period;
+    public final boolean lastInPeriod;
+    public final IPartitioner partitioner;       // Set during (initial) construction and not modifiable via Transformer
```
so is it there just for validation that all nodes has the same partitioner?

jacek-lewandowski commented on src/java/org/apache/cassandra/tcm/listeners/SchemaListener.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.tcm.ClusterMetadata;
+
+public class SchemaListener implements ChangeListener
```
Consider renaming this class to something more like `AlterSchemaListener`, `AlterSchemaTransformationLister`, `SchemaTransformationListener` to not confuse it with `SchemaChangeListener`

ifesdjeen committed:
Add implementation overview doc

jacek-lewandowski committed:
JAVADOC: Startup and Discovery

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/membership/NodeId.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    @Deprecated
+    public UUID toUUID()
```
Brand new method, deprecated?

blerer commented on src/java/org/apache/cassandra/tcm/extensions/AbstractExtensionValue.java:
```diff
@@ -0,0 +1,80 @@
+    public V withLastModified(Epoch lastModified)
+    {
+        this.lastModified = lastModified;
+        return (V)this;
```
This method is broken for its sub-classes. For example `IntValue` will attempt to cast an `IntValue` object into an `Integer`.
There is also a change in the class expectation outside of `ExtensionValue` classes all others `MetadataValue` seems to be immutable. I think that it would make sense to keep that immutability everywhere.

ifesdjeen committed:
Use pinned Harry version

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/transformations/Register.java:
```diff
@@ -0,0 +1,222 @@
+        return nodeId;
+    }
+
+    private static NodeId register(boolean force)
```
`force` is always `false`?

blerer committed:
add javadoc

ekaterinadimitrova2 committed:
Small JavaDoc, warnings, and code style fixes

ekaterinadimitrova2 committed:
ninja fix: Initialize class - moved braces by mistake

ekaterinadimitrova2 commented on test/distributed/org/apache/cassandra/distributed/test/log/RegisterTest.java:
```diff
@@ -0,0 +1,169 @@
+                try
+                {
+                    Field field = NodeVersion.class.getDeclaredField("CURRENT");
+                    Field modifiers = Field.class.getDeclaredField("modifiers");
```
This will not work with JDK 17 because of JDK-8210522 (rewrite reflection of "modifiers" field)
We fixed other in-jvm tests in [CASSANDRA-18181](https://issues.apache.org/jira/browse/CASSANDRA-18181)

jacek-lewandowski commented on src/java/org/apache/cassandra/cql3/statements/schema/AlterKeyspaceStatement.java:
```diff
@@ -128,14 +142,21 @@ private void validateNoRangeMovements()
-        List<InetAddressAndPort> notNormalEndpoints = endpoints.filter(endpoint -> !FBUtilities.getBroadcastAddressAndPort().equals(endpoint) &&
-                                                                                   !Gossiper.instance.getEndpointStateForEndpoint(endpoint).isNormalState())
-                                                               .collect(Collectors.toList());
+        ClusterMetadata metadata = ClusterMetadata.current();
```
Referring `ClusterMetadata.current()` from the transformation code is wrong, so as referring to the configuration, `DatabaseDescriptor`, `Gossiper` and others. This can be seen in many schema transformations and makes the transformations non-pure. In turn, it can make replaying the log lead to inconsistent metadata for the same epoch on different nodes.

blerer commented on src/java/org/apache/cassandra/tcm/membership/NodeId.java:
```diff
@@ -0,0 +1,120 @@
+    public static boolean isValidNodeId(UUID uuid)
+    {
+        return uuid.getMostSignificantBits() == 0 &&
+               uuid.getLeastSignificantBits() < Integer.MAX_VALUE;
```
It seems that `uuid.getLeastSignificantBits()` can return a negative number that will not fit in an integer but will be smaller than `Integer.MAX_VALUE`.

blerer commented on src/java/org/apache/cassandra/tcm/membership/NodeId.java:
```diff
@@ -0,0 +1,120 @@
+{
+    public static final Serializer serializer = new Serializer();
+
+    private final int id;
```
Looking at the code it is unclear what are the valid values for id an if it should be considered as a signed or unsigned number.

blerer commented on src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java:
```diff
@@ -0,0 +1,280 @@
+            newKeyspaces.forEach(ksm -> {
+               ksm.tables.forEach(tm -> {
+                   if (tm.epoch.isAfter(prev.nextEpoch()))
+                       throw new InvalidRequestException(String.format("Invalid schema transformation. " +
+                                                                       "Resultant epoch for table metadata of %s.%s (%d) " +
+                                                                       "is greater than for cluster metadata (%d)",
+                                                                       ksm.name, tm.name, tm.epoch.getEpoch(),
+                                                                       prev.nextEpoch().getEpoch()));
```
`InvalidRequestException` has always been used for error that were linked to a user request that was trying to do something that C* was not supporting (e.g. ordering by non-clustering columns). In the current patch, there are multiple places where we use that exception for errors on which the user has no control (e.g. some race conditions or internal errors). I believe that we need to change that because it can be really confusing from the user point of view.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/locator/InetAddressAndPort.java:
```diff
@@ -316,6 +328,26 @@ static int getDefaultPort()
 
+    public static final class MetadataSerializer implements org.apache.cassandra.tcm.serialization.MetadataSerializer<InetAddressAndPort>
+    {
+        public static final MetadataSerializer serializer = new MetadataSerializer();
```
```suggestion
        public static final MetadataSerializer SERIALIZER = new MetadataSerializer();
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1926,4 +1969,109 @@ public static TopPartitionTracker.StoredTopPartitions getTopPartitions(TableMeta
+    public static ByteBuffer getSnapshot(Epoch epoch)
+    {
+        logger.info("Getting snapshot of epoch = {}", epoch);
+        String query = String.format("SELECT SNAPSHOT FROM %s.%s WHERE epoch = ?", SchemaConstants.SYSTEM_KEYSPACE_NAME, SNAPSHOT_TABLE_NAME);
```
```suggestion
        String query = String.format("SELECT snapshot FROM %s.%s WHERE epoch = ?", SchemaConstants.SYSTEM_KEYSPACE_NAME, SNAPSHOT_TABLE_NAME);
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/ReplicationParams.java:
```diff
@@ -17,21 +17,40 @@
 {
+    private static final ReplicationParams META = new ReplicationParams(MetaStrategy.class, ImmutableMap.of());
+
+    public static final Serializer serializer = new Serializer();
+    public static final MessageSerializer messageSerializer = new MessageSerializer();
```
```suggestion
    public static final Serializer SERIALIZER = new Serializer();
    public static final MessageSerializer MESSAGE_SERIALIZER = new MessageSerializer();
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/Schema.java:
```diff
@@ -75,235 +57,126 @@
  * production code (would be great if they were not accessed in the test code as well).
  * <p>
- * TL;DR: All modifications are made using the implementation of {@link SchemaUpdateHandler} obtained from the provided
+ * TL;DR: All modifications are made using the implementation of SchemaUpdateHandle obtained from the provided
```
This doc needs update; To begin with - there are no `SchemaUpdateHandle` and `mergeAndUpdateVersion` or do I miss anything?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaKeyspace.java:
```diff
@@ -67,7 +67,7 @@
  *
  * Please notice this class is _not_ thread safe and all methods which reads or updates the data in schema keyspace
- * should be accessed only from the implementation of {@link SchemaUpdateHandler} in synchronized blocks.
+ * should be accessed only from the implementation of {SchemaUpdateHandler} in synchronized blocks.
```
It seems Java doc around Schema needs a re-write at more places. There is no `SchemaUpdateHandler` anymore

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaKeyspace.java:
```diff
@@ -287,7 +287,7 @@ public static KeyspaceMetadata metadata()
     }
 
-    static Collection<Mutation> convertSchemaDiffToMutations(KeyspacesDiff diff, long timestamp)
+    public static Collection<Mutation> convertSchemaDiffToMutations(KeyspacesDiff diff, long timestamp)
```
Does this really needs to be public?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaPullVerbHandler.java:
```diff
@@ -32,6 +32,7 @@
  * Sends it's current schema state in form of mutations in response to the remote node's request.
  * Such a request is made when one of the nodes, by means of Gossip, detects schema disagreement in the ring.
  */
+@Deprecated
```
This will explode on rebase because of [CASSANDRA-18912](https://issues.apache.org/jira/browse/CASSANDRA-18912) We need to add `since` and maybe even a bit more info like why it was deprecated and when it is for removal (if ever)

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaPushVerbHandler.java:
```diff
@@ -35,6 +35,7 @@
  * (which is going to act as coordinator) and that node sends (pushes) it's updated schema state
  * (in form of mutations) to all the alive nodes in the cluster.
  */
+@Deprecated
```
This will explode on rebase because of [CASSANDRA-18912](https://issues.apache.org/jira/browse/CASSANDRA-18912) We need to add `since` and maybe even a bit more info like why it was deprecated and when it is for removal (if ever)

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaTransformation.java:
```diff
@@ -17,20 +17,76 @@
+     *   cluster metadata for the peer.
+     *   </li>
+     * </ul>
+     * At the moment, not all SchemaTransformation are entirely free of side effects, because both client warnings and
+     * guardrails are tightly coupled to the ST implementations and they do produce side effects. Ideally, ClientWarn
+     * and Guardrails can be reworked to decouple them from ClientState and threadlocals. In the meantime, we can hint
+     * to them that SchemaTransformation::apply is being called in the context of AlterSchema::execute and so they
+     * should not produce their side effects. The default impl is a no-op, but AlterSchemaStatement which has access to
+     * a ClientState, is able to pause both warnings and guardrails.
+     * @see org.apache.cassandra.cql3.statements.schema.AlterSchemaStatement#enterExecution()
```
What are the side effects you mean here, except duplicate warnings?
I saw the `org.apache.cassandra.cql3.statements.schema.AlterSchemaStatement#enterExecution()` method.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaTransformations.java:
```diff
@@ -31,6 +35,14 @@
  */
 public class SchemaTransformations
 {
+    public static SchemaTransformation fromCql(String cql)
```
Not used anwyhere

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaTransformations.java:
```diff
@@ -128,8 +143,9 @@ public static SchemaTransformation addTypes(Types toAdd, boolean ignoreIfExists)
      */
     public static SchemaTransformation addView(ViewMetadata view, boolean ignoreIfExists)
```
Not used anywhere

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SchemaUpdateHandler.java:
```diff
@@ -1,78 +0,0 @@
- * The interface has been extracted to abstract out that functionality. It allows for various implementations like
- * Gossip based (the default), ETCD, offline, etc., and make it easier for mocking in unit tests.
- */
-public interface SchemaUpdateHandler
```
Isn't removal of an interface a breaking change?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/SystemDistributedKeyspace.java:
```diff
@@ -228,10 +229,6 @@ public static void successfulParentRepair(TimeUUID parent_id, Collection<Range<T
     {
-        // Don't record repair history if an upgrade is in progress as version 3 nodes generates errors
-        // due to schema differences
-        boolean includeNewColumns = !Gossiper.instance.hasMajorVersion3Nodes();
```
This was leftover from 4? We do not support upgrades from 3 to 5.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -59,25 +61,34 @@
 @Unmetered
 public class TableMetadata implements SchemaElement
 {
+    public static final Serializer serializer = new Serializer();
```
```suggestion
    public static final Serializer SERIALIZER = new Serializer();
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/schema/TableMetadataRef.java:
```diff
@@ -15,33 +15,62 @@
+
 package org.apache.cassandra.schema;
 
-import org.github.jamm.Unmetered;
```
Why `@Unmetered` was removed?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/AtomicLongBackedProcessor.java:
```diff
@@ -0,0 +1,182 @@
+
+    public static class InMemoryMetadataSnapshots implements MetadataSnapshots
+    {
+        private SortedMap<Epoch, ClusterMetadata> snapshots;
```
```suggestion
        private final SortedMap<Epoch, ClusterMetadata> snapshots;
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tcm/Commit.java:
```diff
@@ -0,0 +1,429 @@
+    }
+
+    static volatile Result.Serializer resultSerializerCache;
+    public static interface Result
```
```suggestion
    public interface Result
```

blerer commented on src/java/org/apache/cassandra/tcm/Discovery.java:
```diff
@@ -0,0 +1,249 @@
+        NOT_STARTED,
+        IN_PROGRESS,
+        FINISHED,
+        FOUND_CMS
```
`FOUND_CMS` is not used anymore.

blerer commented on src/java/org/apache/cassandra/tcm/Discovery.java:
```diff
@@ -0,0 +1,249 @@
+    public DiscoveredNodes discover(int rounds)
+    {
+        boolean res = state.compareAndSet(State.NOT_STARTED, State.IN_PROGRESS);
+        assert res : String.format("Can not start discovery as it is in state %s", state.get());
```
The code relies on this assertion to behave correctly. Should it not be a real check?
